pkg/controllers/nodeclaim/tagging/controller.go

```diff
@@ -0,0 +1,136 @@
+	"github.com/aws/karpenter/pkg/providers/instance"
+	"github.com/aws/karpenter/pkg/utils"
+
+	corev1beta1 "github.com/aws/karpenter-core/pkg/apis/v1beta1"
+	"github.com/aws/karpenter-core/pkg/cloudprovider"
+	corecontroller "github.com/aws/karpenter-core/pkg/operator/controller"
+)
+
+const (
+	TagNodeClaim = corev1beta1.Group + "/nodeclaim"
+	TagName      = "Name"
+)
+
+type Controller struct {
+	kubeClient       client.Client
+	instanceProvider *instance.Provider
+}
+
+func NewController(kubeClient client.Client, instanceProvider *instance.Provider) corecontroller.Controller {
+	return corecontroller.Typed[*corev1beta1.NodeClaim](kubeClient, &Controller{
+		kubeClient:       kubeClient,
+		instanceProvider: instanceProvider,
+	})
+}
+
+func (c *Controller) Name() string {
+	return "nodeclaim.tagging"
+}
+
+func (c *Controller) Reconcile(ctx context.Context, nodeClaim *corev1beta1.NodeClaim) (reconcile.Result, error) {
+	storedNodeClaim := nodeClaim.DeepCopy()
+
+	if !isTaggable(nodeClaim) {
+		return reconcile.Result{}, nil
+	}
+
+	id, err := utils.ParseInstanceID(nodeClaim.Status.ProviderID)
+	if err != nil {
+		// We don't throw an error here since we don't want to retry until the ProviderID has been updated.
+		logging.FromContext(ctx).Errorf("failed to parse instance ID, %w", err)
+		return reconcile.Result{}, nil
+	}
+
+	err = c.tagInstance(ctx, nodeClaim, id)
+	if err != nil {
+		if !cloudprovider.IsNodeClaimNotFoundError(err) {
+			return reconcile.Result{}, err
+		}
+		return reconcile.Result{}, nil
+	}
+
+	nodeClaim.Annotations = lo.Assign(nodeClaim.Annotations, map[string]string{v1beta1.AnnotationInstanceTagged: "true"})
+	if !equality.Semantic.DeepEqual(nodeClaim, storedNodeClaim) {
+		if err := c.kubeClient.Patch(ctx, nodeClaim, client.MergeFrom(storedNodeClaim)); err != nil {
+			return reconcile.Result{}, client.IgnoreNotFound(err)
+		}
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func (c *Controller) Builder(_ context.Context, m manager.Manager) corecontroller.Builder {
+	return corecontroller.Adapt(
+		controllerruntime.
+			NewControllerManagedBy(m).
+			For(&corev1beta1.NodeClaim{}).
+			WithEventFilter(predicate.NewPredicateFuncs(func(o client.Object) bool {
+				nc := o.(*corev1beta1.NodeClaim)
+				return isTaggable(nc)
+			})),
+	)
+}
+
+func (c *Controller) tagInstance(ctx context.Context, nc *corev1beta1.NodeClaim, id string) error {
+	tags := map[string]string{
+		TagName:      nc.Status.NodeName,
+		TagNodeClaim: nc.Name,
+	}
+
+	// Remove tags which have been already populated
+	instance, err := c.instanceProvider.Get(ctx, id)
+	if err != nil {
+		return err
+	}
+	tags = lo.OmitByKeys(tags, lo.Keys(instance.Tags))
+
+	if len(tags) != 0 {
+		err = c.instanceProvider.CreateTags(ctx, id, tags)
+		time.Sleep(time.Second)
+		return err
+	}
+	return nil
+}
+
+func isTaggable(nc *corev1beta1.NodeClaim) bool {
+	// Instance has already been tagged
+	if val := nc.Annotations[v1beta1.AnnotationInstanceTagged]; val == "true" {
```

Comment:
Is it possible that we'll want to dynamically tag for a nodeClaim multiple times in the future? For instance, let's say we want to make sure that instance tags sync up with a specific set of values, I'm assuming we'll probably have to remove this check on the annotation, and then probably remove the annotation altogether. I'd rather not add an annotation if we know we're going to remove it anyway.

Reply:
That seems reasonable to me. The intention behind using an annotation rather than tracking in memory was so it would persist through controller restarts. In a large scale cluster this will result in a high number of `DescribeInstance` calls right at startup but I don't know if that would actually be an issue.

Reply:
Our DescribeInstances is [batched](https://github.com/aws/karpenter/blob/d1d1371ae2e1552b8fdded7d343bf24ea18bee31/pkg/batcher/describeinstances.go#L36), so I don't think the number of calls here would be a problem, if anything it would be a problem to have a large queue depth, but that's an issue with other controllers too.

Reply:
Aligned offline to keep the annotation to minimize API calls upon controller restart as well as minimize the time new requests have to wait after restart.